Feature/tigoncli

gokulavasan opened the pull request:
- Adding a Command Line Interface for Tigon Distributed Mode.
- Adding ability to announce services in Flows and discover it using FlowControllers.
- Adding basic logback.xml files to suppress chatty libraries.
- Add sequencing to FlowTwillApplication - start txManager TwillRunnable first so that it is the last one to shutdown (Twill guarantees that ordering)

Commit:
Starting TxManager first and the rest of the Flowlet Instances since stopping of flowlets is done in the reverse order

Commit:
CLI for distributed

Commit:
Ability to change instances of Flowlets in distributed mode

Commit:
We can handle lib dir in JARs

Commit:
Merging develop

Commit:
Merge remote-tracking branch 'origin/develop' into feature/tigoncli

Commit:
Java docs

Commit:
Improve error reporting in CLI

Commit:
Adding sequencing to the order in which twill runnables are started in TwillApplication - txManager first and Flowlets later which will make sure txManager is shutdown last

Commit:
txManager does not need programjar

Commit:
Merge remote-tracking branch 'origin/develop' into feature/tigoncli

gokulavasan commented:
Bamboo shows GREEN. 
https://bamboo-prod.continuuity.com/browse/PRODRELALL-UNIT9-2

sreevatsanraman commented:
Overall design wise - we should make this similar to CDAP cli. Interms of feature - we should add a PrefixCompleter. We can do these in a follow-up PR after the 9/25. Please create a JIRA to track these.

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/DistributedMain.java:
```diff
@@ -103,24 +106,33 @@ public static void main(String[] args) {
       }
 
-      if (args.length != 2) {
+      if (args.length < 2) {
```
If zookeeper quorum is not passed, isn't it supposed to take from the HConf?

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/DistributedMain.java:
```diff
@@ -133,24 +145,62 @@ public void run() {
-    addShutDownHook();
-    controller = deployClient.deployFlow(jarPath, mainClassName, jarUnpackDir);
-    runLatch.await();
+  public void startUp(PrintStream out) throws Exception {
```
As mentioned in the other comment. We need to revise this similar to Cdap Cli

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+
+  START (2, "Starts a Flow", "<Path-to-Jar> <FlowClassName>"),
+  LIST (0, "Lists all Flows which are currently running", ""),
+  STOP (1, "Stops a Flow", "<FlowName>"),
```
Status command?

awholegunch replied:
<flow-name>

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/cli/DistributedFlowOperations.java:
```diff
@@ -0,0 +1,247 @@
+    Iterable<TwillController> controllers = lookupFlow(flowName);
+    for (TwillController controller : controllers) {
+      Iterable<Discoverable> iterable = controller.discoverService(service);
+      sleepForZK();
```
This seems very brittle. Do we need this sleep?

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/DistributedMain.java:
```diff
@@ -133,24 +145,62 @@ public void run() {
+    consoleReader.setPrompt("tigon> ");
+    String line;
+    while ((line = consoleReader.readLine()) != null) {
+      String[] args = line.split(" ");
```
check if args is not null and args.length >= 1

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/DistributedMain.java:
```diff
@@ -133,24 +145,62 @@ public void run() {
+    while ((line = consoleReader.readLine()) != null) {
+      String[] args = line.split(" ");
+      String command = args[0].toUpperCase();
+      CLICommands cmd = null;
```
This can be moved inside try block

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/DistributedMain.java:
```diff
@@ -133,24 +145,62 @@ public void run() {
+          throw new InvalidCLIArgumentException(cmd.printHelp());
+        }
+        if (cmd.equals(CLICommands.START)) {
+          flowOperations.startFlow(new File(args[1]), args[2]);
```
Starts will be async. So there is no indication to user that the flow has started? it would be good to give some feedback to the user that the operation has been successful or failed. Understand if we want to clean this up in a new PR after the lauch.

sreevatsanraman replied:
Similar comment for stop, delete, setinstances

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+  DELETE (1, "Stops and Deletes the Queues for a Flow", "<FlowName>"),
+  SERVICEINFO (1, "Prints all Services announced in a Flow", "<FlowName>"),
+  DISCOVER (2, "Discovers a service endpoint for a Flow", "<FlowName> <ServiceName>"),
+  SET (3, "Set Flowlet instances for a Flow", "<FlowName> <FlowletName> <Instances>"),
```
Set the number of instance of a Flowlet for a Flow

awholegunch replied:
"<flow-name> <flowlet-name> <instances>"

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+  STOP (1, "Stops a Flow", "<FlowName>"),
+  DELETE (1, "Stops and Deletes the Queues for a Flow", "<FlowName>"),
+  SERVICEINFO (1, "Prints all Services announced in a Flow", "<FlowName>"),
+  DISCOVER (2, "Discovers a service endpoint for a Flow", "<FlowName> <ServiceName>"),
```
Discovers a Service endpoint for a Flow

awholegunch replied:
"<flow-name><service-name>"

awholegunch replied:
Should this be "Discovers a service endpoint of a Service <service-name> for a Flow <flow-name>"?

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public enum CLICommands {
+
+  START (2, "Starts a Flow", "<Path-to-Jar> <FlowClassName>"),
```
I think we have been using all lowercase to date as descriptions of parameters:
"<path-to-jar> <flow-classname>"

awholegunch replied:
Also, don't we need to pass the name the Flow is to be set to? How do you refer to it later?

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+  START (2, "Starts a Flow", "<Path-to-Jar> <FlowClassName>"),
+  LIST (0, "Lists all Flows which are currently running", ""),
+  STOP (1, "Stops a Flow", "<FlowName>"),
+  DELETE (1, "Stops and Deletes the Queues for a Flow", "<FlowName>"),
```
<flow-name>

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+  SERVICEINFO (1, "Prints all Services announced in a Flow", "<FlowName>"),
+  DISCOVER (2, "Discovers a service endpoint for a Flow", "<FlowName> <ServiceName>"),
+  SET (3, "Set Flowlet instances for a Flow", "<FlowName> <FlowletName> <Instances>"),
+  FLOWLETINFO (1, "Prints Flowlet Names and corresponding Instances", "<FlowName>"),
```
"Prints Flowlet Names and corresponding Instances for a Flow", "<flow-name>"

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+  DISCOVER (2, "Discovers a service endpoint for a Flow", "<FlowName> <ServiceName>"),
+  SET (3, "Set Flowlet instances for a Flow", "<FlowName> <FlowletName> <Instances>"),
+  FLOWLETINFO (1, "Prints Flowlet Names and corresponding Instances", "<FlowName>"),
+  SHOWLOGS (1, "Shows live logs for the Flow", "<FlowName>"),
```
"Shows live logs for a Flow", "<flow-name>"

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+  LIST (0, "Lists all Flows which are currently running", ""),
+  STOP (1, "Stops a Flow", "<FlowName>"),
+  DELETE (1, "Stops and Deletes the Queues for a Flow", "<FlowName>"),
+  SERVICEINFO (1, "Prints all Services announced in a Flow", "<FlowName>"),
```
"<flow-name>"

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+  SET (3, "Set Flowlet instances for a Flow", "<FlowName> <FlowletName> <Instances>"),
+  FLOWLETINFO (1, "Prints Flowlet Names and corresponding Instances", "<FlowName>"),
+  SHOWLOGS (1, "Shows live logs for the Flow", "<FlowName>"),
+  QUIT (0, "Quit Tigon Client", "");
```
"Quit the Tigon Client"

awholegunch replied:
In CDAP CLI, this is "exit".

awholegunch commented on tigon-client/src/main/java/co/cask/tigon/cli/CLICommands.java:
```diff
@@ -0,0 +1,53 @@
+    return argCount + 1;
+  }
+
+  public String printHelp() {
```
In CDAP CLI, this is "help".
There is also a version command.

Commit:
Fix classloading problem and address some PR comments

Commit:
Pass userargs

Commit:
Use fileutils

Commit:
Test fix

Commit:
Fix classloading in distributed mode

Commit:
PR comments

Commit:
Pretty print

gokulavasan commented:
@sreevatsanraman @awholegunch Addressed your comments. Please review. 
logbacks need to fixed for distributed mode. but since many PRs are awaiting changes in this, I would like to get this merged sooner than later.

gokulavasan commented:
Bamboo is GREEN. https://bamboo-prod.continuuity.com/browse/PRODRELALL-UNIT9

Commit:
Split by any whitespace

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/cli/DistributedFlowOperations.java:
```diff
@@ -0,0 +1,265 @@
+  @Override
+  public List<InetSocketAddress> discover(String flowName, String service) {
+    List<InetSocketAddress> endPoints = Lists.newArrayList();
+    Iterable<TwillController> controllers = lookupFlow(flowName);
```
This can be null right? We need to handle.

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/cli/DistributedFlowOperations.java:
```diff
@@ -0,0 +1,265 @@
+
+  @Override
+  public void stopFlow(String flowName) {
+    Iterable<TwillController> controllers = lookupFlow(flowName);
```
Same here.

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/cli/DistributedFlowOperations.java:
```diff
@@ -0,0 +1,265 @@
+    try {
+      for (int i = 0; i < count; i++) {
+        if (obj != null) {
+          TimeUnit.MILLISECONDS.sleep(250);
```
Why sleep if you found an object.

sreevatsanraman commented on tigon-client/src/main/java/co/cask/tigon/cli/DistributedFlowOperations.java:
```diff
@@ -0,0 +1,265 @@
+    return iterable;
+  }
+
+  private void sleepForZK(Object obj) {
```
Let's discuss about this function in person.

sreevatsanraman commented on tigon-flow/src/main/java/co/cask/tigon/internal/app/runtime/distributed/FlowTwillApplication.java:
```diff
@@ -84,12 +89,16 @@ public TwillSpecification configure() {
+    String firstFlowlet = flowletNames.remove(0);
+    String[] flowletArray = new String[flowletNames.size()];
+    flowletArray = flowletNames.toArray(flowletArray);
+    return runnableSetter.withOrder().begin(TXMANAGERNAME).nextWhenStarted(firstFlowlet, flowletArray)
```
You don't need a new flowletArray variable. It could be simplified as

``` java
flowletNames.subList(..).toArray()
```

Commit:
Addressing PR comments

gokulavasan commented:
@sreevatsanraman Can you take a look again? Addressed your comments.

sreevatsanraman commented:
Let's Open Jira in OSS for the improvements we want to make to CLI client
a) Design change to be like CDAP CLI
b) Handling the logs 
c) Prefix helper

Add it as TODO in CLI and merge.

gokulavasan commented:
@sreevatsanraman Thanks Sree. Opened Jiras in OSS. Merging now.
